### DISCUSSION TITLE
Support to fix "error 5" when compiling WW3

### DISCUSSION THREAD

**Discussion by Cesar-IIUNAM** [NOAA-EMC/WW3#1066] [Q&A] [answered]
**Hello everyone! best regards.**
**Please, I need your support to compile the model. I have tried several ways but I have not been able to compile the model on my** **desktop computer.** 

**I have tried the following:**

wget https://github.com/NOAA-EMC/WW3/archive/refs/tags/6.07.tar.1.gz
tar -zxf 6.07.1.tar.gz
cp comp.Gnu comp
cp link.Gnu link
cp switch_Ifremer2 switch
./w3_setup ../
./w3_make

**When I do ./w3_make I get the following error:**
(base) ctorresm@pcctorres:/media/ctorresm/Datos/CROCO/WW3-6.07.1/model/bin$ ./w3_make
 
                *********************************
              ***   WAVEWATCH III make      ***
                *****************************
 
 Exporting WWATCH3_ENV=/media/ctorresm/Datos/CROCO/WW3-6.07.1/model/bin/wwatch3.env
Main directory    : /media/ctorresm/Datos/CROCO/WW3-6.07.1/model
Scratch directory : /media/ctorresm/Datos/CROCO/WW3-6.07.1/model/tmp
Save source codes : yes
Save listings     : yes
 
[INFO] list of the programs selected : ww3_grid ww3_strt ww3_bound ww3_outf ww3_outp ww3_trck ww3_grib ww3_gint gx_outf gx_outp ww3_uprstr ww3_gspl ww3_prep ww3_shel ww3_prnc ww3_ounf ww3_ounp ww3_bounc ww3_trnc ww3_prtide ww3_multi ww3_sbs1 libww3 ww3_systrk
 
Processing ww3_grid
---------------------
ad3 : processing constants 
      2 errors found, see /media/ctorresm/Datos/CROCO/WW3-6.07.1/model/tmp/constants.err
      *** compile error ***
          file constants.o not found
make: *** [makefile:844: /media/ctorresm/Datos/CROCO/WW3-6.07.1/model/obj/constants.o] Error 5

**My variables are defined like this:**

export CC=gcc
export FC=gfortran
export F90=gfortran
export F77=gfortran

#Librerias MPI
export CC=mpicc
export FC=mpifort
export F90=mpifort
export F77=mpifort
export MPIF90=mpifort

export WWATCH3_NETCDF=NC4
export NETCDF_CONFIG=/usr/bin/nc-config

#export NETCDF_CONFIG=/usr/bin
#export NETCDF_ROOT=/usr

**when I do nc-config --all and nf-config --all, I get:**

**nf-config --all**

This netCDF-Fortran 4.5.4 has been built with the following features: 

  --cc        -> gcc
  --cflags    -> -I/usr/include -Wdate-time -D_FORTIFY_SOURCE=2

  --fc        -> gfortran
  --fflags    -> -I/usr/include -I/usr/include
  --flibs     -> -L/usr/lib/x86_64-linux-gnu -lnetcdff -Wl,-Bsymbolic-functions -flto=auto -ffat-lto-objects -flto=auto -Wl,-z,relro -Wl,-z,now -lnetcdf -lnetcdf -lm 
  --has-f90   -> 
  --has-f03   -> yes

  --has-nc2   -> yes
  --has-nc4   -> yes

  --prefix    -> /usr
  --includedir-> /usr/include
  --version   -> netCDF-Fortran 4.5.4

**nc-config --all**

This netCDF 4.8.1 has been built with the following features: 

  --cc            -> /usr/bin/cc
  --cflags        -> -I/usr/include -I/usr/include/hdf5/serial
  --libs          -> -L/usr/lib/x86_64-linux-gnu -L/usr/lib/x86_64-linux-gnu/hdf5/serial -lnetcdf
  --static        -> -lhdf5_hl -lhdf5 -lcrypto -lcurl -lpthread -lsz -lz -ldl -lm

  --has-c++       -> no
  --cxx           -> 

  --has-c++4      -> no
  --cxx4          -> 

  --has-fortran   -> yes
  --fc            -> gfortran
  --fflags        -> -I/usr/include -I/usr/include
  --flibs         -> -L/usr/lib/x86_64-linux-gnu -lnetcdff -Wl,-Bsymbolic-functions -flto=auto -ffat-lto-objects -flto=auto -Wl,-z,relro -Wl,-z,now -lnetcdf -lnetcdf -lm
  --has-f90       -> 
  --has-f03       -> yes

  --has-dap       -> yes
  --has-dap2      -> yes
  --has-dap4      -> yes
  --has-nc2       -> yes
  --has-nc4       -> yes
  --has-hdf5      -> yes
  --has-hdf4      -> no
  --has-logging   -> no
  --has-pnetcdf   -> no
  --has-szlib     -> yes
  --has-cdf5      -> yes
  --has-parallel4 -> no
  --has-parallel  -> no
  --has-nczarr    -> yes

  --prefix        -> /usr
  --includedir    -> /usr/include
  --libdir        -> /usr/lib/x86_64-linux-gnu
  --version       -> netCDF 4.8.1

**At the moment, this is the information I am sharing. If you need anything else, do not hesitate to ask for it. Please, any suggestions** **would be greatly appreciated.**

**Best regards,**
**Cesar.**

**Comment by mickaelaccensi** [comment]
Hi Cesar,
To know the exact error message, you need to open the log file : /media/ctorresm/Datos/CROCO/WW3-6.07.1/model/tmp/constants.err

I really recommend you to move to the latest release of the develop branch of WW3 and to use cmake to build WW3, it is more robust and user friendly.

There is a tutorial about ww3 installation that we do during our annual training course:
https://docs.google.com/document/d/1azvnWsozWDpSo6c7gSFwZVAsqQ5wcpLndsXDXe1W4gY/edit?usp=sharing

**Comment by Cesar-IIUNAM** [comment]
Hi Mickael,
Thank for your kind reply and for sharing de training material.
I have tried several ways to install and compile the model on my computer. Perhaps with your tutorial I can do it.

As for the contents of the log file, it says the following:

gfortran: error: unrecognized command line option ‘-fallow-invalid-boz’
gfortran: error: unrecognized command line option ‘-fallow-argument-mismatch’; did you mean ‘-Wno-argument-mismatch’?

Best regards,
César.

**Reply by mickaelaccensi** [reply]
the error is due to the fixed compiler option in the comp.gnu and link.gnu, that's why using cmake is much more convenient to avoid this kind of problem

**Reply by Cesar-IIUNAM** [reply]
Mickael, thanks for your answer. I had already tried this method. I have seen this error before. 
I have checked my environment variable settings and I don't see anything unusual. I have the latest version of cmake installed.

According to the guide you shared, I have tried:
 cd $HOME/WW3
 rm -rf build 
 mkdir build
 cd build
 cmake .. -DSWITCH=IINGEN_st4   (my computer is desktop!)

By the way, my switch file I have configured as follows: 
SHRD NC4 PR3 UQ ST4 STAB0 FLX0 LN1 NL1 IC0 IS0 REF0 BT1 DB1 MLIM TR0 BS0 WNX1 WNT1 CRX1 CRT1 O0 O1 O2 O4 O5 O6 O7

I get this error:

<img width="1076" alt="image" src="https://github.com/NOAA-EMC/WW3/assets/113042121/edd425e6-6fb2-4ecd-b201-96568dbc1e2c">

I also attached screenshot of the first part of CMakeError.log (it seems it is not possible to attach a file). I guess the error has to do with the compiler, but I don't know what else to do about it. Any support would be greatly appreciated.

<img width="1108" alt="image" src="https://github.com/NOAA-EMC/WW3/assets/113042121/db10e03a-a74e-4719-852f-ba3591947a6c">


best regards,
Cesar.

**Reply by MatthewMasarik-NOAA** [reply]
These errors still point to issues with the compiler.  Compiler and MPI need to be available (loaded if need be) before running the cmake command.  Then run cmake and it will find what is available to it.  When it attempts to do this in your first screen shot, the last line before the text turns red, it is saying that the check for a C compiler failed.

**Reply by MatthewMasarik-NOAA** [reply]
After resolving the compiler, one thing that may help is providing the full path to the switch file, like this  `-DSWITCH=/your/path/to/IINGEN_st4`.

**Reply by Cesar-IIUNAM** [reply]
Matthew,
thanks for your reply, I don't exactly know how to load the compiler or make it available on my desktop. 
On my system (~/.bashrc) I have gfortran defined as compiler, like this:

export CC=gcc
export FC=gfortran
export F90=gfortran
export F77=gfortran

However, in the first screenshot I sent, in the first lines before it turns red it seems to look for "mpicc", I don't understand this.

-César.

**Reply by MatthewMasarik-NOAA** [reply]
Hi César,
the `mpicc` refers to an MPI wrapper for C compiler.  You'll need to install an MPI library if you have not already.  Two free options are MPICH and OpenMPI.  You'll need to look into how to install that for your particular environment since a number of factors may affect that process.

**Reply by mickaelaccensi** [reply]
@MatthewMasarik-NOAA , we should be able to install WW3 without MPI library, it was possible before cmake. 
We probably need to set an environment variable like :
export MPICC=gcc
@Cesar-IIUNAM can you try this also ?

**Reply by Cesar-IIUNAM** [reply]
Hi @mickaelaccensi, in fact I just tried it and I get the same error. Likewise, I installed the openMPI library (I think it shouldn't be necessary since I use a desktop computer), getting the same error.
I honestly don't understand why I'm having these problems.

<img width="1079" alt="image" src="https://github.com/NOAA-EMC/WW3/assets/113042121/789352c0-8c4a-43b5-8c6a-a13603af99d7">

**Reply by mickaelaccensi** [reply]
looking back to your environment variables, you have overwritten the CC and FC variables : 
```
export CC=gcc
export FC=gfortran
export F90=gfortran
export F77=gfortran

#Librerias MPI
export CC=mpicc
export FC=mpifort
export F90=mpifort
export F77=mpifort
export MPIF90=mpifort
```

Also, do you have the exe /usr/bin/mpicc ?

**Reply by Cesar-IIUNAM** [reply]
I had already corrected the definition of the variables for CC and FC. Now I define them like this:

export CC=gcc
export FC=gfortran
export F90=gfortran
export F77=gfortran
export MPICC=gcc

About if i have the exe /usr/bin/mpicc, it's not clear to me what I should check, but doing "$ which mpicc", I get the following: 
/usr/bin/mpicc
Although, perhaps, it's not what you want me to review. Can you tell me exactly what I should check? 

Thank very much for your kind support,
César.

**Reply by MatthewMasarik-NOAA** [reply]
César,
please try this following call (the only thing to change is the path to the switch file):

```
cmake -DCMAKE_Fortran_COMPILER=gfortran -DCMAKE_C_COMPILER=gcc -DSWITCH=/path/to/switchfile ..    
```

**Comment by mickaelaccensi** [comment]
the official installation procedure is described in the readme : 
https://github.com/NOAA-EMC/WW3/blob/develop/model/README.md

**Reply by CarstenHansen** [reply]
Thank you @mickaelaccensi for disclosing this location of the CMake Build information, and for the details (in the Ifremer tutorial) about how to make a local installation of CMake. Now I realized that the WW3 build requires CMake 3.19+ to work, and it actually appears to work smoothly.

**Comment by MatthewMasarik-NOAA** [comment, marked as answer]
Hi @Cesar-IIUNAM,
I wanted to circle back and see if you had a chance to try this latest `cmake` call:
```
cmake -DCMAKE_Fortran_COMPILER=gfortran -DCMAKE_C_COMPILER=gcc -DSWITCH=/path/to/switchfile ..    
```
I think it should resolve the error you displayed at the C compiler check.

**Reply by Cesar-IIUNAM** [reply]
Hi @MatthewMasarik-NOAA, thank very much, and apologies for not responding faster.
I tried and now I have this error. I was also not clear if I should put ".." at the end.  

<img width="851" alt="image" src="https://github.com/NOAA-EMC/WW3/assets/113042121/aeb9b9eb-4a18-40ec-9f28-12eb789e825c">

**Reply by MatthewMasarik-NOAA** [reply]
@Cesar-IIUNAM, that's great -- progress.  Regarding the ` .. ` at the end, yes, those need to be included.  You're calling cmake from `../WW3/build/`, so the two dot specify the directory location one level up, `../WW3/` as the root of the source to be built.  In the first screenshot you are doing it correctly.

So, in that first screenshot you got passed the cmake compiler check, and now this new error is with the switch file.  It is telling you that the switch you are using, `switch_IINGEN_st4`, is not set up to support grib, so I would look into the switches in your file.

**Reply by Cesar-IIUNAM** [reply]
Hi @MatthewMasarik-NOAA , 
Thanks for your support! I have changed my switch and at least now I don't get errors anymore. After "cmake -DCMAKE_Fortran_COMPILER=gfortran -DCMAKE_C_COMPILER=gcc -DSWITCH=/media/ctorresm/Datos/CROCO/WW3/model/bin/switch_IINGEN_st4 .." 

I get the following, is it ok?:
<img width="737" alt="image" src="https://github.com/NOAA-EMC/WW3/assets/113042121/ce09cd19-7441-4529-8f77-48efe45c8ee4">

Based on the guide (https://github.com/NOAA-EMC/WW3/blob/develop/model/README.md), from the build directory, I did: 
"make" 
"make install"  
Which does not seem to be correct. Then I ran it in one level up, ../WW3/ 
Well I am not doing something right yet... 

<img width="544" alt="image" src="https://github.com/NOAA-EMC/WW3/assets/113042121/6ee39aad-985c-4401-b74f-b33987b58cec">

Again, thank very much for your support!

César.

**Reply by MatthewMasarik-NOAA** [reply]
Hi @Cesar-IIUNAM, sure thing.  I'm glad you're making progress.  The error above looks like in the install stage, it is trying to write to a system location that your user does not have permissions for.  You can fix this by adding an additional option to the cmake call:
```
-DCMAKE_INSTALL_PREFIX=<path/to/install/dir>
```
use any path that your user can write to.

**Reply by Cesar-IIUNAM** [reply]
@MatthewMasarik-NOAA, thank you very much, I was able to execute the commands without problems. It seems that the model was installed and compiled. 
What caught my attention is that the /model/exe directory was not created with the programs like "ww3_grid", "ww3_strt" and others. Instead I have seen that in the model directory there is a directory with these programs. I guess it is due to this new version of WW3.

I can only say thank for your support and patience!

<img width="591" alt="image" src="https://github.com/NOAA-EMC/WW3/assets/113042121/3850ae69-6d4c-4d70-a8fa-ac532eccf8ad">

**Reply by MatthewMasarik-NOAA** [reply]
> @MatthewMasarik-NOAA, thank you very much, I was able to execute the commands without problems. It seems that the model was installed and compiled.

@Cesar-IIUNAM , that's great, and you're welcome!

>What caught my attention is that the /model/exe directory was not created with the programs like "ww3_grid", "ww3_strt" and others. Instead I have seen that in the model directory there is a directory with these programs. I guess it is due to this new version of WW3.

That's correct.  cmake allows (and encourages) building projects outside of the model source code area.  It looks like you used the WW3/ dir as the INSTALL_PREFIX.  This is OK, though I tend to specify a sub-directory of WW3/, for instance WW3/install, that will put the build output in during the install step.  In this case, you don't need to manually create the 'install' sub-dir, cmake will do that for you.  

At any rate that's just a preference, and the important part is you got over the compile hurdle!

**Comment by MatthewMasarik-NOAA** [comment]
@Cesar-IIUNAM  also, if this has been addressed to your satisfaction, please feel free to close the discussion thread.  Cheers

**Comment by aronroland** [comment]
Hi,

 

… u need to run make in the build directory … and u can chose the binaries by saying … e.g. make ww3_shel, u can compile in parallel by using make -j 4 ww3_shel … let us know!

 

Aron

 

Von: Cesar-IIUNAM ***@***.***> 
Gesendet: Dienstag, 19. September 2023 19:55
An: NOAA-EMC/WW3 ***@***.***>
Cc: Subscribed ***@***.***>
Betreff: Re: [NOAA-EMC/WW3] Support to fix "error 5" when compiling WW3 (Discussion #1066)

 

Hi @MatthewMasarik-NOAA <https://github.com/MatthewMasarik-NOAA>  ,
Thanks for your support! I have changed my switch and at least now I don't get errors anymore. After "cmake -DCMAKE_Fortran_COMPILER=gfortran -DCMAKE_C_COMPILER=gcc -DSWITCH=/media/ctorresm/Datos/CROCO/WW3/model/bin/switch_IINGEN_st4 .."

I get the following, is it ok?:
 <https://user-images.githubusercontent.com/113042121/269043212-ce09cd19-7441-4529-8f77-48efe45c8ee4.png> 

Based on the guide (https://github.com/NOAA-EMC/WW3/blob/develop/model/README.md), from the build directory, I did:
"make"
"make install"
Which does not seem to be correct. Then I ran it in one level up, ../WW3/
Well I am not doing something right yet...

 <https://user-images.githubusercontent.com/113042121/269049670-6ee39aad-985c-4401-b74f-b33987b58cec.png> 

Again, thank very much for your support!

César.
